x-pack/plugins/cases/public/components/actions/severity/use_severity_action.tsx

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback } from 'react';
+import { EuiContextMenuPanelItemDescriptor } from '@elastic/eui';
+import { CaseSeverity } from '../../../../common/api';
+import { useUpdateCases } from '../../../containers/use_bulk_update_case';
+import { Case } from '../../../../common';
+
+import * as i18n from './translations';
+import { UseActionProps } from '../types';
+import { useCasesContext } from '../../cases_context/use_cases_context';
+import { severities } from '../../severity/config';
+
+const getSeverityToasterMessage = (severity: CaseSeverity, cases: Case[]): string => {
+  const totalCases = cases.length;
+  const caseTitle = totalCases === 1 ? cases[0].title : '';
+
+  switch (severity) {
+    case CaseSeverity.LOW:
+      return i18n.SET_SEVERITY_LOW({ totalCases, caseTitle });
+    case CaseSeverity.MEDIUM:
+      return i18n.SET_SEVERITY_MEDIUM({ totalCases, caseTitle });
+    case CaseSeverity.HIGH:
+      return i18n.SET_SEVERITY_HIGH({ totalCases, caseTitle });
+    case CaseSeverity.CRITICAL:
+      return i18n.SET_SEVERITY_CRITICAL({ totalCases, caseTitle });
+
+    default:
+      return '';
+  }
+};
+
+interface UseSeverityActionProps extends UseActionProps {
+  selectedSeverity?: CaseSeverity;
+}
+
+const shouldDisableSeverity = (cases: Case[], severity: CaseSeverity) =>
+  cases.every((theCase) => theCase.severity === severity);
+
+export const useSeverityAction = ({
+  onAction,
+  onActionSuccess,
+  isDisabled,
+  selectedSeverity,
+}: UseSeverityActionProps) => {
+  const { mutate: updateCases } = useUpdateCases();
+  const { permissions } = useCasesContext();
+  const canUpdateSeverity = permissions.update;
+  const isActionDisabled = isDisabled || !canUpdateSeverity;
+
+  const handleUpdateCaseSeverity = useCallback(
+    (selectedCases: Case[], severity: CaseSeverity) => {
+      onAction();
+      const casesToUpdate = selectedCases.map((theCase) => ({
+        severity,
+        id: theCase.id,
+        version: theCase.version,
+      }));
+
+      updateCases(
+        {
+          cases: casesToUpdate,
+          successToasterTitle: getSeverityToasterMessage(severity, selectedCases),
+        },
+        { onSuccess: onActionSuccess }
+      );
+    },
+    [onAction, updateCases, onActionSuccess]
+  );
+
+  const getSeverityIcon = (severity: CaseSeverity): string =>
+    selectedSeverity && selectedSeverity === severity ? 'check' : 'empty';
+
+  const getActions = (selectedCases: Case[]): EuiContextMenuPanelItemDescriptor[] => {
+    return [
+      {
+        name: severities[CaseSeverity.LOW].label,
+        icon: getSeverityIcon(CaseSeverity.LOW),
+        onClick: () => handleUpdateCaseSeverity(selectedCases, CaseSeverity.LOW),
```

Comment:
I need to get the selected cases somehow which are passed to `getActions`. It invalidates memorization but I think it is fine for this case. I am open to suggestions 🙂.

Reply:
oh I see, yeah that's fine. Or we could use two `useCallback` I think:

`onClick: useCallback(() => handle..., [...])`

Or does that not work?

Reply:
This will not work because the `onClick` handler passes a `MouseEvent` and not the selected cases. Somehow I need to get them from the `getActions` function. I could 

```
import { memoize } from 'lodash'

const onLowSeverityClick = memorize((selectedCases) => handleUpdateCaseSeverity(selectedCases, CaseSeverity.LOW))

//

onClick: () => onLowSeverityClick(selectedCases)
```

but I do not think it is worth the optimization until we have a strong indicator that it causes performance problems. The memorization cost in this scenario maybe be greater than the problem itself.